### ISSUE TITLE
Tensor parallel for convolutions and groupnorm 

### ISSUE THREAD

**Issue by richardk53** [pytorch/pytorch#133221]
### 🚀 The feature, motivation and pitch

Recurrent LLM architectures such as Mamba (v1 and v2) and xLSTM use depthwise convolution and GroupNorm in each block. Both are currently not supported by Pytorch tensor parallel. Depthwise convolution is independent over the channels and GroupNorm corresponds to independent LayerNorm for every head within a block. Therefore, both should be easily parallelizable over the independent heads. However, the current tensor parallel API in Pytorch does not support convolutions and GroupNorm. 

Is this feature currently being developed? Are there fundamental reasons in pytorch distributed that make an extension of Rowwise and Columnwise strategies to convolution and other modules difficult?

### Alternatives

Megatron supports the depthwise separable convolution and the groupnorm for mamba, but it is not based on the pytorch tensor parallel API.

**Comment by tianyu-l** [comment]
I agree these are features TP should support. Are you interested in submitting PRs by improving [style.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/style.py) so we can help review; o/w we may need to wait until we have bandwidth to do it.

**Comment by richardk53** [comment]
@tianyu-l I will need to look into this in the next weeks for sure. But at the moment, I am unsure whether to use [style.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/style.py) or avoid it entirely and instead build on a framework such as Megatron. 
I am wondering about the roadmap, do you think you will continue with [style.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/style.py) or replace this by a different approach soon? I am suspicious that it might not be continued, because no other modules are mentioned in the documentation, I could find a roadmap, and torch.distributed.tensor does not receive commits recently. 
I also just found https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_tp_conv.py#L106. Where is this TP for convolutions used currently? It does not seem to be used in torch.distributed.tensor. So I am wondering if a different API is being developed instead...

**Comment by tianyu-l** [comment]
style.py is the place to go!

- pytorch Tensor Parallel is only recently beta-released (https://pytorch.org/docs/stable/distributed.tensor.parallel.html). We will continue working on it. I think the reason TP is only supported on selected module types (e.g. Linear, Embedding, Norm, etc.) is because our focus has been on llama-style model. I agree we should continue make it better, and we appreciate your feedback and potential contribution.
- There is no other effort planned for TP.
- DTensor (torch.distributed.tensor) is the backbone of our TP. DTensor itself is almost ready and will be beta-released shortly. We are constantly developing it.
- The _tp_conv.py you linked is used here (https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L100). It's for low-level DTensor treatment of the op `aten.convolution.default`, and is not directly related to the high-level TP API.

**Comment by nimit** [comment]
@tianyu-l
I am trying to TP nn.Conv1D by using distribute_module and trying to split the input tensor colwise (Shard(1) in distribute_tensor)
This leads to an error because in [_conv_ops.py](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_conv_ops.py), the convolution rules seem to be made for Conv2D and expect (N, C, W, H) as input shapes.
Am I correct in assuming that I need to create new rules for Conv1D?
How can I link them to the conv1d function call?

**Comment by tianyu-l** [comment]
@nimit 
> Am I correct in assuming that I need to create new rules for Conv1D?

Yes, I think so. You'll need to create a similar rule / strategy for the conv1d aten ops (forward and backward).

To find what aten ops to support, the error msg should contain it, e.g. https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_sharding_prop.py#L456

E.g. the corresponding aten op for forward Conv1D might be
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L1711

Let me know if you need further help.

**Comment by tiffany940107** [comment]
Hi @tianyu-l , I wonder do you have a Dtensor support for GroupNorm? I see groupnorm op forward function does not have mean and variance in its inputs, so I wonder how to update mean and variance of different ranks by Dtensor? Do I have to modify groupnorm implementation like "layer_norm_with_mean_rstd_out" here? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/layer_norm.cpp#L40C13-L40C42

**Comment by tianyu-l** [comment]
@tiffany940107 you'll need to add DTensor sharding propagation support for the group norm op in pytorch
F.group_norm: https://pytorch.org/docs/stable/generated/torch.nn.functional.group_norm.html
aten level op: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L2982

It will look similar to what DTensor does for layer norm
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_math_ops.py#L797

**Comment by xingchensong** [comment]
@tianyu-l  hi, does PyTorch have plans to natively support tp for conv1d/group norm?

**Comment by tianyu-l** [comment]
@xingchensong Yeah supporting it natively sounds good, although we haven't found the bandwidth to work on it yet. Are you interested in contributing?

**Comment by xingchensong** [comment]
> @xingchensong Yeah supporting it natively sounds good, although we haven't found the bandwidth to work on it yet. Are you interested in contributing?

Yep, I'm happy to submit a PR when I solve this problem ~

**Comment by tiffany940107** [comment]
Hi @tianyu-l , are you planning to support conv3d parallel？I'm trying to use Dtensor to support this op, as now AIGC tasks need conv3d essentially.
